File: hasPathSum.java
package edu.ucsc;


/**
 * Leetcode Number:112
 * Problem Name:Path Sum
 *
 * @author: Created by Xiaotong Li 
 * @time: 6/20/18:2018-06-20 18:50.
 */
public class hasPathSum {
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) {
            return false;
        }
        if (root.left == null && root.right == null && sum - root.val == 0) {
            return true;
        }
        return hasPathSum(root.left, sum - root.val) || hasPathSum(root.right, sum - root.val);
    }
}

File: Python/BinaryTree/hasPathSum.py
#! /Users/xiaotongli/anaconda3/bin/python
# -*- coding: utf-8 -*-
# @Time    : 10/12/18 10:26 PM
# @Author  : Xiaotong Li
# @School  : University of California, Santa Cruz
# @FileName: hasPathSum.py
# @Software: PyCharm


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        if not root:
            return False
        stack = [(root, root.val)]
        while stack:
            curr, val = stack.pop()
            if not curr.right and not curr.left:
                if val == sum:
                    return True
            if curr.right:
                stack.append((curr.right, val + curr.right.val))
            if curr.left:
                stack.append((curr.left, val + curr.left.val))
        return False
